cedar-lean/Cedar/Spec/PartialExpr.lean

```diff
@@ -0,0 +1,212 @@
+/-
+ Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Data
+import Cedar.Spec.Expr
+import Cedar.Thm.Data.Map -- necessary for a termination argument
+import Cedar.Thm.Termination.Expr
+import Cedar.Thm.Termination.Value
+
+/-! This file defines abstract syntax for Cedar expressions. -/
+
+namespace Cedar.Spec
+
+open Cedar.Data
+
+/--
+  Identical to `Expr` except that it has an `unknown` case, and the recursive
+  elements are also all `PartialExpr` instead of `Expr`
+-/
+inductive PartialExpr where
+  | lit (p : Prim)
+  | var (v : Var)
+  | ite (cond : PartialExpr) (thenExpr : PartialExpr) (elseExpr : PartialExpr)
+  | and (a : PartialExpr) (b : PartialExpr)
+  | or (a : PartialExpr) (b : PartialExpr)
+  | unaryApp (op : UnaryOp) (expr : PartialExpr)
+  | binaryApp (op : BinaryOp) (a : PartialExpr) (b : PartialExpr)
+  | getAttr (expr : PartialExpr) (attr : Attr)
+  | hasAttr (expr : PartialExpr) (attr : Attr)
+  | set (ls : List PartialExpr)
+  | record (map : List (Attr × PartialExpr))
+  | call (xfn : ExtFun) (args : List PartialExpr)
+  | unknown (name : String)
+
+deriving instance Repr, Inhabited for PartialExpr
+
+def Value.asPartialExpr (v : Value) : PartialExpr :=
+  match v with
+  | .prim p => .lit p
+  | .set s => .set (s.elts.map₁ λ ⟨v, h₁⟩ =>
+      have := Value.set_termination v s h₁
+      v.asPartialExpr)
+  | .record m => .record (m.kvs.map₁ λ ⟨(k, v), h₁⟩ =>
+      have := Value.record_termination v m (Map.in_kvs_values h₁)
+      (k, v.asPartialExpr))
+  | .ext (.decimal d) => .call ExtFun.decimal [PartialExpr.lit (.string d.unParse)]
+  | .ext (.ipaddr ip) => .call ExtFun.ip [PartialExpr.lit (.string (Cedar.Spec.Ext.IPAddr.unParse ip))]
+termination_by v.size
+
+/--
+  A version of `PartialExpr`, but only allows "restricted expressions" -- no
+  vars, no expressions that require entity data to evaluate, no operators at all,
+  just literals, unknowns, extension values, and sets/records of those things
+-/
+inductive RestrictedPartialExpr where
```

Comment:
Curious:  why do we need `RestrictedPartialExpr`s?

Depending on what they are used for (e.g., just proofs?), it may be easier to define an inductive predicate that restricts `PartialExpr`s rather than defining a whole new type (and presumably translating back and forth between regular and restricted partial exprs).

Reply:
They are used as the residual type for `RestrictedPartialValue`.  Obviously, that leads to the question "why do we need `RestrictedPartialValue`.

`RestrictedPartialValue` is the type of attributes in a `PartialEntityData` and in the `context` of a `PartialRequest`.  It is clear that we cannot allow fully general `PartialValue` in these locations, because that would allow entity/context attribute values to be residual expressions that depend on the values of other entity/context attributes, or of the variables `principal`, `action`, `resource`.  However, you make a good suggestion that we could simply require some well-formedness predicate on `PartialEntityData` and `PartialRequest` that requires that any residuals must only use one of the allowed constructors/variants.  (Indeed my proofs already need well-formedness predicates on these, eg ensuring that the underlying `Map`s are `WellFormed`, so it would be easy to add another requirement to the definitions of well-formedness there.)

Similarly, `RestrictedPartialValue` is the type allowed in a substitution-for-unknowns operation -- when providing the value for an `unknown`, you can provide any `RestrictedPartialValue`, but it clearly does not make sense to allow providing a fully general `PartialValue` and thus allowing the value of the `unknown` to depend on Cedar residual expressions involving entity or request data.  However, this could be expressed as a precondition instead of at the type level.

Reply:
I guess the higher-level design question is, in situations like this when we need to establish constraints on the inputs users provide (in this case, user-provided partial values for attribute values, context values, and unknown substitutions), do we prefer expressing this at the type level (so any value of the given type is valid), or via preconditions on the top-level methods.

Reply:
It depends on a couple of things.  Here is my rough mental checklist for deciding on one vs the other:

(1) Is expressing this at the type level making the specification clearer to a reader?
(2) Is expressing this at the type level making the algorithm (in this case PE) shorter / easier to read / etc. in some way?  For example, does having this type help you return `.ok` (vs `.error`) in more places?
(3) Does the algorithm make sense without the type-level constraint?

These are pretty subjective, but if the answer is "yes"-ish for 1-2 and "no"-ish for 3, then I keep it at the type level.  Otherwise, it becomes a well-formedness (WF) predicate. 

For example, in the symbolic compilation algorithm, we use WF predicates because (a) the WF constraints are subtle and cannot be expressed as ADT types (we'd have to use dependent types), and (b) the algorithm does _something_ on any input, and we prove that it does what we want on WF inputs.

Reply:
I think that regardless of the answer in the full context of existing and planned PE proofs, it's certainly clear that the need for the `Restricted*` types is not justified in the current PR.  So I will take them out here, and if/when they are needed, we can reopen this discussion in the future PR where we have a concrete justification.

Reply:
Removed from the current PR.